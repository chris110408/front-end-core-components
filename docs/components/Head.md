# Head

The Head component uses [react-helmet-async](https://www.npmjs.com/package/react-helmet-async) to render head tags. It's based on the premise that there are 3 incoming sources of head tag data.

* Site level head data
* Page level head data
* Component level head data

In ascending order of importance, site, page and component data is merged and duplicates removed to form the final set of head tags. For example if the meta description tag is set within the site head data, but then also set at the page level head data, then the value from the site level is overwritten by the page level.

Assume the following:

```javascript
const site = {
  title: 'My website',
  meta: [
    { name: 'description', content: 'A description of my website' },
    { name: 'og:description', content: 'Open graph description' }
  ]
}

const page = {
  title: 'My page on my website',
  meta: [
    { name: 'description', content: 'A description of my page on my website' }
  ]
}

const component = {
  meta: [
    { name: 'description', content: 'A description of a specific page' }
  ]
}
```

will result in the output:

```html
<head>
  ...
  <!-- Page level overwites site level title -->
  <title>My page on my website</title>
  <!-- Component level overwrites page and site level description meta tag -->
  <meta name="description" content="A description of a specific page" />
  <!-- Open graph (og:) tags are transformed automatically to use property instead of name -->
  <meta property="og:description" content="Open graph description">
  ...
</head>
```

## Dependencies

* react `^16.3`
* prop-types `^15.6.2`
* react-helmet-async `^1.0.4`

## Usage

```javascript
import React from 'react'
import { Head } from '@input-output-hk/front-end-core-components'

const MyApp = () => (
  <div>
    <Head
      // Site level data
      site={{
        title: 'My website',
        meta: [
          { name: 'description' content: 'My websites description' }
        ]
      }}
      // Page level data
      page={{
        title: 'My websites page',
        meta: [
          { name: 'description' content: 'My websites page description' }
        ]
      }}
      // Component level data
      component={{
        title: 'My websites pages component',
        meta: [
          { name: 'description' content: 'My websites pages component description' }
        ]
      }}
      // Optional language for the site
      lang='en'
      // Optional locale for the website
      locale='en-US'
      // Optional list of available locales for your website
      availableLocales={[ 'en-US', 'fr-FR' ]}
      // Full URL used for meta data which requires full URL's such as Twitter tags
      // and open graph tags
      // Remove any trailing slash here
      url='https://mywebsite.com'
    />
    <body>
      ...
    </body>
  </div>
)

```

The example above is to show how the props can be passed to the `Head` component. At IOHK we use the [Gatsby](https://www.gatsbyjs.org/) framework, hence why we have split the head data into 3 distinct levels of priority. The meta data we use comes from markdown files and is transformed at build time to be consumed by the `Head` component.
