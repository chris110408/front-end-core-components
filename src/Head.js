import React from 'react'
import PropTypes from 'prop-types'
import { Helmet } from 'react-helmet-async'
import { isRelative } from './lib/url'

const ALLOW_DUPLICATE_TAGS = [ 'og:locale:alternate' ]

/**
 * Generates og:locale:alternative head tags from available locales
 *
 * @param {Array<String>} availableLocales
 * @param {String} locale
 * @returns {Array<Object>}
 */
function getGeneratedHeadData (availableLocales, locale) {
  const alternativeLocales = [ ...availableLocales ]
  alternativeLocales.splice(availableLocales.indexOf(locale), 1)
  return alternativeLocales.map(locale => ({
    name: 'og:locale:alternate',
    content: locale
  }))
}

/**
 * Builds the head tags from three sources from highest to lowest priority i.e.
 * component overwrites duplicate page or site meta data and so on and so forth
 *
 *  - component level meta data
 *  - page level meta data
 *  - site level meta data
 *
 *
 * @param {Object} params
 * @param {Object} params.componentHeadData
 * @param {Object} params.pageHeadData
 * @param {Object} params.siteHeadData
 * @param {Array<String>} params.availableLocales
 * @param {String} params.locale
 * @param {String} params.url
 * @returns {Object} object with title and complete meta array
 */
function getHeadData ({ component, page, site, availableLocales, locale, url }) {
  const filteredPage = {}
  Object.keys(page).forEach(key => {
    if (page[key]) filteredPage[key] = page[key]
  })

  const componentMeta = component.meta || []
  if (component.title) {
    componentMeta.push({ name: 'twitter:title', content: component.title })
    componentMeta.push({ name: 'og:title', content: component.title })
  }

  const metaNames = []
  const meta = []
  const unfilteredMeta = [
    ...(getGeneratedHeadData(availableLocales, locale)),
    ...componentMeta,
    ...(filteredPage.meta || []),
    ...(site.meta || [])
  ]

  const getFileURL = file => isRelative(file) ? `${url}${file}` : file
  unfilteredMeta.forEach((tag) => {
    if (!tag.content && !tag.file) return
    if (ALLOW_DUPLICATE_TAGS.includes(tag.name) || !metaNames.includes(tag.name)) meta.push({ name: tag.name, content: tag.content || getFileURL(tag.file) })
    metaNames.push(tag.name)
  })

  return {
    title: componentMeta.title || filteredPage.title || site.title,
    meta: meta.map(tag => {
      if (tag.name.match(/^og:/)) return { property: tag.name, content: tag.content }
      return tag
    })
  }
}

const Head = ({
  site: { title: siteTitle = '', meta: siteMeta = [] } = {},
  page: { title: pageTitle = '', meta: pageMeta = [] } = {},
  component: { title: componentTitle = '', meta: componentMeta = [] } = {},
  locale,
  availableLocales = [],
  lang,
  url,
  children
}) => {
  const completeHeadData = getHeadData({
    component: { title: componentTitle, meta: componentMeta },
    page: { title: pageTitle, meta: pageMeta },
    site: { title: siteTitle, meta: siteMeta },
    availableLocales,
    locale,
    url
  })

  return (
    <Helmet
      title={completeHeadData.title}
      meta={completeHeadData.meta}
    >
      {lang && <html lang={lang} />}
      {children}
    </Helmet>
  )
}

Head.propTypes = {
  site: PropTypes.shape({
    title: PropTypes.string,
    meta: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      content: PropTypes.string,
      file: PropTypes.string
    }))
  }),
  page: PropTypes.shape({
    title: PropTypes.string,
    meta: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      content: PropTypes.string,
      file: PropTypes.string
    }))
  }),
  component: PropTypes.shape({
    title: PropTypes.string,
    meta: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      content: PropTypes.string,
      file: PropTypes.string
    }))
  }),
  locale: PropTypes.string,
  availableLocales: PropTypes.arrayOf(PropTypes.string),
  lang: PropTypes.string,
  url: (props, propName, componentName) => {
    if (typeof props[propName] !== 'string') return new Error(`'${propName}' prop for component '${componentName}' must be a string`)
    if (!props[propName].match(/^(https?:)\/\/[^/]+$/)) {
      return new Error(
        `'${propName}' prop for component '${componentName}' must be a valid URL starting with "http://", "https://" or "//" with no trailing slash, e.g. "https://example.org" or "http://localhost:8080"`
      )
    }
  },
  children: PropTypes.node
}

export default Head
