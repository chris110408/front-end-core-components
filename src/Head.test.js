import React from 'react'
import Head from './Head'
import { shallow } from 'enzyme'

const DEFAULT_PROPS = {
  url: 'https://example.iohk.io'
}

function getWrapper (props = {}) {
  return shallow(
    <Head
      {...DEFAULT_PROPS}
      {...props}
    />
  )
}

describe('<Head />', () => {
  test('it renders correctly with minimal props', () => {
    expect(getWrapper()).toMatchSnapshot()
  })

  test('it renders lang correctly', () => {
    expect(getWrapper({ lang: 'en' })).toMatchSnapshot()
  })

  test('it renders locales correctly', () => {
    expect(getWrapper({ locale: 'en-US', availableLocales: [ 'en-GB', 'en-US', 'fr-FR' ] })).toMatchSnapshot()
  })

  test('it renders children components correctly', () => {
    expect(getWrapper({ children: (
      <html lang='en' />
    ) })).toMatchSnapshot()
  })

  test('it renders tags based on levels correctly', () => {
    const props = {
      site: {
        title: 'The website title',
        meta: [
          { name: 'description', content: 'My website description' },
          { name: 'twitter:title', content: 'Share on Twitter!' },
          { name: 'twitter:image', file: '/relative/image.jpeg' },
          { name: 'twitter:image:alt', content: 'My image' },
          { name: 'og:title', content: 'OG' },
          { name: 'og:image', file: 'https://www.images.com/image.jpeg' }
        ]
      },
      page: {
        title: 'Page title',
        meta: [
          { name: 'description', content: 'Page description' },
          { name: 'keywords', content: 'website,site' }
        ]
      },
      component: {
        meta: [
          { name: 'og:title', content: 'Component OG title' },
          { name: 'keywords', content: 'my,website,site' }
        ]
      },
      url: 'https://www.images.com'
    }

    expect(getWrapper(props)).toMatchSnapshot()
  })
})
