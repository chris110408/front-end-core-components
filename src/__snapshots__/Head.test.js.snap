// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Head /> it renders children components correctly 1`] = `
<Helmet
  defer={true}
  encodeSpecialCharacters={true}
  meta={Array []}
  title=""
>
  <html
    lang="en"
  />
</Helmet>
`;

exports[`<Head /> it renders correctly with minimal props 1`] = `
<Helmet
  defer={true}
  encodeSpecialCharacters={true}
  meta={Array []}
  title=""
/>
`;

exports[`<Head /> it renders lang correctly 1`] = `
<Helmet
  defer={true}
  encodeSpecialCharacters={true}
  meta={Array []}
  title=""
>
  <html
    lang="en"
  />
</Helmet>
`;

exports[`<Head /> it renders locales correctly 1`] = `
<Helmet
  defer={true}
  encodeSpecialCharacters={true}
  meta={
    Array [
      Object {
        "content": "en-GB",
        "property": "og:locale:alternate",
      },
      Object {
        "content": "fr-FR",
        "property": "og:locale:alternate",
      },
    ]
  }
  title=""
/>
`;

exports[`<Head /> it renders tags based on levels correctly 1`] = `
<Helmet
  defer={true}
  encodeSpecialCharacters={true}
  meta={
    Array [
      Object {
        "content": "Component OG title",
        "property": "og:title",
      },
      Object {
        "content": "my,website,site",
        "name": "keywords",
      },
      Object {
        "content": "Page description",
        "name": "description",
      },
      Object {
        "content": "Share on Twitter!",
        "name": "twitter:title",
      },
      Object {
        "content": "https://www.images.com/relative/image.jpeg",
        "name": "twitter:image",
      },
      Object {
        "content": "My image",
        "name": "twitter:image:alt",
      },
      Object {
        "content": "https://www.images.com/image.jpeg",
        "property": "og:image",
      },
    ]
  }
  title="Page title"
/>
`;
